allennlp/commands/train.py

```diff
@@ -418,23 +453,24 @@ def _train_worker(
         batch_weight_key=batch_weight_key,
     )
 
-    try:
-        if distributed:  # let the setup get ready for all the workers
-            dist.barrier()
-
-        metrics = train_loop.run()
-    except KeyboardInterrupt:
-        # if we have completed an epoch, try to create a model archive.
-        if master and os.path.exists(os.path.join(serialization_dir, _DEFAULT_WEIGHTS)):
-            logging.info(
-                "Training interrupted by the user. Attempting to create "
-                "a model archive using the current best epoch weights."
-            )
-            archive_model(serialization_dir)
-        raise
-
-    if master:
-        train_loop.finish(metrics)
+    if not dry_run:
```

Comment:
If you're removing `dry-run`, I expected your recommendation to be to just use a no-op trainer, instead of adding an argument here.  I'm not sure I'm a fan of adding the argument, especially in a place like this block, where the logic gets quite nested.

In general, jack-of-all-trades functions are not good software design, because they lead to spaghetti code.  Merging `fine-tune` and `train` made sense because there was so much duplication, and when I actually removed the fine-tune command, there were no extra parameters needed anywhere (in fact you removed the ones that were added previously).  The amount of duplication with `dry-run` is much smaller (because the code in `dry-run` is much smaller), and I'm not convinced that the small reduction in duplication is worth the added complexity here.

@DeNeutoy, any opinion?

Reply:
See the 1st commit as a reference of how `dry-run` looks refactored out. I could change it to use a no-op trainer if you want.

Reply:
The refactoring in that commit seems like a strict improvement over what was there before, and it would be pretty obvious to accept a PR that just did that.  It's the further combining it with `train` that I'm less inclined about.

Reply:
Yeah, I understand. From a practical point of view, a dry run should be like a normal run but without running. If they're kept separate, then it's not gonna be an actual run (e.g., there can be changes in `train` that are by mistake not made in `dry-run`).